### DISCUSSION TITLE
Connection timeout working is not same as described in the clientBuilder[Bug]

### DISCUSSION THREAD

**Discussion by ragaur-tibco** [apache/pulsar#20344] [Q&A]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

Java Client(2.11.0)-->broker(2.11.0)
Broker instance OS: Amazon Linux 2023
Java App running OS : RHEL or Ubuntu or Win 10

### Minimal reproduce step

client = PulsarClient.builder()
						.serviceUrl("pulsar://broker.example.com:6650/")
						.connectionTimeout(30, TimeUnit.SECONDS)
						.build();

### What did you expect to see?

The duration of time to wait for a connection to be established with a broker has been set to 30 seconds. In the event that the duration passes without a response from the broker, the connection attempt should be dropped within the specified time frame.

### What did you see instead?

We have set the duration of time to 30 sec wait for a connection to a broker to be established. But the duration passes without a response from the broker, the connection attempt is dropped is taking more time as mentioned or configured.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by tisonkun** [comment]
@rgaur12 can you elaborate a bit what's the real duration the connection attempt is dropped? Also can you share the log that you identify the issue as you described?

I go through the code and it simply passes the config option to Netty's bootstrap so that the logic should be correct:

```java
// ConnectionPool.java
bootstrap.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, conf.getConnectionTimeoutMs());
```

**Reply by ragaur-tibco** [reply]
creating the pulsar connection with invalid host IP/Broker URL which doesn't exist and connection timeout 10 seconds.
While building the client it is taking more then 10 sec to throw exception then given time in connection timeout and retrying to connect to broker.

` <CausedBy> TIBCO-BW-CORE-500051: Activity [PulsarProducer] fault.   <CausedBy> com.tibco.bw.palette.pulsar.runtime.fault.PulsarPluginException: TIBCO-BW-PALETTE-PULSAR-500008: Exception occurred while sending a message [java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/{"previous":[{"attempt":0,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":1,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":2,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":3,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":4,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":5,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":6,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":7,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":8,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"},{"attempt":9,"error":"java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: pulsar://broker.example.com:6650/"}]}].-{ActivityName=PulsarProducer, ProcessName=pulsarsrtest.Process, ModuleName=PulsarSRtest}`
